plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/OpenLineageListener.java

```diff
@@ -0,0 +1,340 @@
+            throws IllegalAccessException
+    {
+        boolean failed = queryCompletedEvent.getMetadata().getQueryState().equals("FAILED");
+
+        RunFacetsBuilder runFacetsBuilder = getBaseRunFacetsBuilder(queryCompletedEvent.getContext());
+
+        runFacetsBuilder.put(OpenLineageTrinoFacet.TRINO_METADATA.getText(),
+                getTrinoMetadataFacet(queryCompletedEvent.getMetadata()));
+        runFacetsBuilder.put(OpenLineageTrinoFacet.TRINO_QUERY_CONTEXT.getText(),
+                getTrinoQueryContextFacet(queryCompletedEvent.getContext()));
+        runFacetsBuilder.put(OpenLineageTrinoFacet.TRINO_QUERY_STATISTICS.getText(),
+                getTrinoQueryStatisticsFacet(queryCompletedEvent.getStatistics()));
+
+        if (failed) {
+            queryCompletedEvent
+                    .getFailureInfo()
+                    .flatMap(QueryFailureInfo::getFailureMessage)
+                    .ifPresent(failureMessage -> runFacetsBuilder
+                            .errorMessage(openLineage
+                                    .newErrorMessageRunFacetBuilder()
+                                    .message(failureMessage)
+                                    .build()));
+        }
+
+        return openLineage.newRunEventBuilder()
+                .eventType(
+                        failed
+                                ? RunEvent.EventType.FAIL
+                                : RunEvent.EventType.COMPLETE)
+                .eventTime(queryCompletedEvent.getEndTime().atZone(ZoneId.of("UTC")))
+                .run(openLineage.newRunBuilder().runId(runID).facets(runFacetsBuilder.build()).build())
+                .job(getBaseJobBuilder(queryCompletedEvent.getMetadata()).build())
+                .inputs(buildInputs(queryCompletedEvent.getIoMetadata()))
+                .outputs(buildOutputs(queryCompletedEvent.getIoMetadata()))
+                .build();
+    }
+
+    private RunFacetsBuilder getBaseRunFacetsBuilder(QueryContext queryContext)
+    {
+        return openLineage.newRunFacetsBuilder()
+                .processing_engine(openLineage.newProcessingEngineRunFacetBuilder()
+                        .name("trino")
+                        .version(queryContext.getServerVersion())
+                        .build());
+    }
+
+    private JobBuilder getBaseJobBuilder(QueryMetadata queryMetadata)
+    {
+        return openLineage.newJobBuilder()
+                .namespace(this.namespace)
+                .name(queryMetadata.getQueryId())
+                .facets(openLineage.newJobFacetsBuilder()
+                            .jobType(openLineage.newJobTypeJobFacet("BATCH", "TRINO", "QUERY"))
+                            .sql(openLineage.newSQLJobFacet(queryMetadata.getQuery()))
+                            .build());
+    }
+
+    private List<InputDataset> buildInputs(QueryIOMetadata ioMetadata)
+    {
+        return ioMetadata.getInputs().stream().map(inputMetadata ->
```

Comment:
To re-iterate on the same question - using `QueryIOMetadata#getInputs` doesn't recognize views and considers only normal table - so for a `CTAS` query `CREATE TABLE new_table AS SELECT * FROM view_1` where SQL under represented by view_1 would be `SELECT * FROM table_1` and the lineage would point to from`new_table -> table_1` and view information is potentially lost in this operation - This is not a recommended operation right ? For those types of query we need to map the the lineage between `new_table` and `view_1`